### PR TITLE
[SPARK-23841][ML] NodeIdCache should unpersist the last cached nodeIdsForInstances

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?
unpersist the last cached nodeIdsForInstances in `deleteAllCheckpoints`

## How was this patch tested?
existing tests

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala [hunk 1]
```diff
@@ -95,7 +95,7 @@ private[spark] class NodeIdCache(
       splits: Array[Array[Split]]): Unit = {
     if (prevNodeIdsForInstances != null) {
       // Unpersist the previous one if one exists.
-      prevNodeIdsForInstances.unpersist()
+      prevNodeIdsForInstances.unpersist(false)
```
Comment (sujithjay) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala]: Is this change required? Should not the call to `unpersist` remain blocking?

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala [hunk 2]
```diff
@@ -166,9 +166,13 @@ private[spark] class NodeIdCache(
...
       }
     }
+    if (nodeIdsForInstances != null) {
+      // Unpersist current one if one exists.
+      nodeIdsForInstances.unpersist(false)
+    }
     if (prevNodeIdsForInstances != null) {
       // Unpersist the previous one if one exists.
-      prevNodeIdsForInstances.unpersist()
+      prevNodeIdsForInstances.unpersist(false)
```
Comment (sujithjay) [inline mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala]: Same question as above. `deleteAllCheckpoints` is blocking because it involves calls to `FileSystem.delete`. So, does it make sense to make the call to `unpersist` non-blocking? Am I missing something here?

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala (hunk 1 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala]: This is not required, but it is usually safe to unpersist without blocking in MLLIB.

File: mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala (hunk 2 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/tree/impl/NodeIdCache.scala]: For now `deleteAllCheckpoints` is only called once in whole MLLIB, and current `unpsersit` of `prevNodeIdsForInstances` is in it. So I think we do not need to impl another method to unpersist datasets (like `PeriodicCheckpointer`)

**Comment by zhengruifeng** [comment]
@srowen  Could you please help reviewing this? Thanks in advance

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master
